msg/msg.go

```diff
@@ -91,3 +40,136 @@ func CreateMsg(metrics []*schema.MetricData, id int64, version Format) ([]byte,
+		}
+		metrics = []*schema.MetricData(out)
+	case FormatMetricDataV1ArrayJson:
+		oldMetrics := make([]*schemaV1.MetricData, 0)
+		err := json.Unmarshal(b[9:], &oldMetrics)
+		if err != nil {
+			return nil, fmt.Errorf("ERROR: failure to unmarshal message body via format %q: %s", format, err)
+		}
+		metrics = make([]schema.MetricData, len(oldMetrics))
+		for i, om := range oldMetrics {
+			metrics[i] = &schema.MetricData{
+				MetricMetadata: schema.MetricMetadata{
+					Id:    om.Id,
+					OrgId: om.OrgId,
+					Name:  om.Name,
+					Unit:  om.Unit,
+					Mtype: om.Mtype,
+					Tags:  om.Tags,
+				},
+				Point: schema.Point{
+					Ts:  uint32(om.Time),
+					Val: om.Value,
+				},
+			}
+		}
+	case FormatMetricDataV1ArrayMsgp:
+		oldMetrics := make([]*schemaV1.MetricData, 0)
+		out := schemaV1.MetricDataArray(oldMetrics)
+		_, err := out.UnmarshalMsg(b[9:])
+		if err != nil {
+			return nil, fmt.Errorf("ERROR: failure to unmarshal message body via format %q: %s", format, err)
+		}
+		oldMetrics = []*schemaV1.MetricData(out)
+		for i, om := range oldMetrics {
+			metrics[i] = &schema.MetricData{
+				MetricMetadata: schema.MetricMetadata{
+					Id:    om.Id,
+					OrgId: om.OrgId,
+					Name:  om.Name,
+					Unit:  om.Unit,
+					Mtype: om.Mtype,
+					Tags:  om.Tags,
+				},
+				Point: schema.Point{
+					Ts:  uint32(om.Time),
+					Val: om.Value,
+				},
+			}
+		}
+	default:
+		return nil, fmt.Errorf(errFmtUnknownFormat, format)
+	}
+	return metrics, nil
+}
+
+// Marshal a DataPoint to byte stream
+func EncodeDataPoint(p schema.DataPoint, b []byte) ([]byte, error) {
+	buf := bytes.NewBuffer(b)
+	switch p.(type) {
+	case *schema.MetricData:
+		err := binary.Write(buf, binary.LittleEndian, uint8(FormatMetricData))
+		if err != nil {
+			return nil, fmt.Errorf(errFmtBinWriteFailed, err)
+		}
+		msg, err := p.(*schema.MetricData).MarshalMsg(nil)
+		if err != nil {
+			return nil, fmt.Errorf(errFmtBinWriteFailed, err)
+		}
+		buf.Write(msg)
+	case *schema.MetricPoint:
+		err := binary.Write(buf, binary.LittleEndian, uint8(FormatMetricPoint))
+		if err != nil {
+			return nil, fmt.Errorf(errFmtBinWriteFailed, err)
+		}
+		msg, err := p.(*schema.MetricPoint).MarshalMsg(nil)
+		if err != nil {
+			return nil, fmt.Errorf(errFmtBinWriteFailed, err)
+		}
+		buf.Write(msg)
+	default:
+		return nil, fmt.Errorf("unknown DataPoint type")
+	}
+
+	return buf.Bytes(), nil
+}
+
+func DecodeDataPoint(b []byte) (schema.DataPoint, error) {
+	switch Format(b[0]) {
+	case FormatMetricPoint:
+		p := &schema.MetricPoint{}
+		_, err := p.UnmarshalMsg(b[1:])
+		if err != nil {
+			return nil, err
+		}
+		return p, nil
+	case FormatMetricData:
+		p := &schema.MetricData{}
+		_, err := p.UnmarshalMsg(b[1:])
+		if err != nil {
+			return nil, err
+		}
+		return p, nil
+	default:
+		// if format is not known, lets assume that this is a MetricData message with no version byte
```

Comment:
by handling it like this, isn't there a risk that we miss a potential error case and end up inserting garbage data into the backend store because the received metrics were somehow corrupted and we didn't notice it due to this `default` case?

Reply:
if the only new message type is the compact payload i proposed in https://github.com/grafana/metrictank/pull/849, then we can also rely on the length of the message for additional confidence.  (the proposed payload is exactly 28 bytes)